### PR TITLE
Include the Log Out option when a custom menu is used

### LINKED ISSUE #644: "LogOut" options missing in the default all-in-one cr config
`Log Out` menu option is missing in the default CR file.
This file provided by jaeger-operator.

```yaml
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: my-jaeger
  namespace: jkandasa
spec:
  strategy: allInOne
  allInOne:
    image: 'registry.redhat.io/distributed-tracing/jaeger-all-in-one-rhel7:1.13'
    options:
      log-level: debug
      query:
        base-path: /jaeger
  ui:
    options:
      dependencies:
        menuEnabled: false
      tracking:
        gaID: UA-000000-2
      menu:
        - label: About Jaeger
          items:
            - label: Documentation
              url: 'https://www.jaegertracing.io/docs/latest'
  storage:
    options:
      memory:
        max-traces: 100000
```

image

### PR DESCRIPTION
**Pull request by jpkrohling**
Closes #644 by automatically adding a "Log Out" option when one seems to be missing. Note that this is only relevant when the OAuth Proxy is being used, which is currently the case only for OpenShift deployments.

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
@jkandasa would you be able to give this one a try ?

**Comment by jkandasa** [comment]
@jpkrohling I tested this operator(quay.io/jpkroehling/jaeger-operator:644-Add-Logout). I can see the `Log Out` option.

image


jaeger operator:
```
time="2020-01-21T10:04:29Z" level=info msg=Versions arch=amd64 identity=openshift-operators.jaeger-operator jaeger=1.16.0 jaeger-operator=v1.16.0-16-g9bb95df3 operator-sdk=v0.12.0 os=linux version=go1.13.4
```

CR Files used:
```yaml
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaegerqe
  namespace: jkandasa
spec:
  strategy: allInOne
  allInOne:
    image: 'registry.redhat.io/distributed-tracing/jaeger-all-in-one-rhel7:1.13'
    options:
      log-level: debug
      query:
        base-path: /jaeger
  ui:
    options:
      dependencies:
        menuEnabled: false
      tracking:
        gaID: UA-000000-2
      menu:
        - label: About Jaeger
          items:
            - label: Documentation
              url: 'https://www.jaegertracing.io/docs/latest'
  storage:
    options:
      memory:
        max-traces: 100000
```

When I add the option (security: none) on CR, Log Out options is not displayed.
```
  ingress:
    security: none
```

**Comment by jpkrohling** [comment]
@pavolloffay would you be able to review this one?

File: pkg/strategy/controller.go [hunk 1]
```diff
@@ -233,40 +233,64 @@ func disableDependenciesTab(uiOpts map[string]interface{}, storage string, depsE
...
+		// build the URL. If needed, this can be a list of patterns in the future
+		if strings.Contains(url, "/oauth/sign_in") {
+			return
+		}
+	}
 
-	json.Unmarshal([]byte(menuStr), &menuArray)
+	logout := map[string]interface{}{
+		"label":        "Log Out",
+		"url":          "/oauth/sign_in",
```
Comment (pavolloffay) [inline pkg/strategy/controller.go]: `sign_in` for log out?

File: pkg/apis/jaegertracing/v1/jaeger_types.go [hunk 2]
```diff
@@ -261,6 +261,10 @@ type JaegerIngressOpenShiftSpec struct {
 
 	// +optional
 	HtpasswdFile string `json:"htpasswdFile,omitempty"`
+
+	// SkipLogout tells the operator to not automatically add a "Log Out" menu option to the custom Jaeger configuration
+	// +optional
+	SkipLogout *bool `json:"skipLogout,omitempty"`
```
Comment (pavolloffay) [inline pkg/apis/jaegertracing/v1/jaeger_types.go]: does it need to be a pointer? We probably want the default to be false.

File: pkg/strategy/controller.go (hunk 1 above)
Comment (jpkrohling) [reply pkg/strategy/controller.go]: Yes :)

File: pkg/apis/jaegertracing/v1/jaeger_types.go (hunk 2 above)
Comment (jpkrohling) [reply pkg/apis/jaegertracing/v1/jaeger_types.go]: For all new properties, we probably want a tri-state value: `nil` (not set), `true`, `false`. This way, the property doesn't get persisted as part of the CR, and we can have different defaults depending on the platform.

**Pull request merged**